rllab/dynamics_randomization/variation.py

```diff
@@ -0,0 +1,204 @@
+
+    @property
+    def distribution(self):
+        return self._distribution
+
+    @distribution.setter
+    def distribution(self, distribution):
+        self._distribution = distribution
+
+    @property
+    def var_range(self):
+        return self._var_range
+
+    @var_range.setter
+    def var_range(self, var_range):
+        self._var_range = var_range
+
+
+class Variations:
+    """
+    The purpose of this class is to keep a list of all the variations
+    that have to be applied to the RandomizedEnv class.
+    The class implements the fluent interface pattern, so each call
+    to set an attribute will return the instance of this class.
+    """
+
+    def __init__(self):
+        self._list = []
+
+    def randomize(self):
+        """
+        Creates a new entry in the list of variations. After calling this
+        method, call the setters for each of the attributes to be used with
+        this new entry using the fluent interface pattern.
+        """
+        variation = Variation()
+        self._list.append(variation)
+        return self
+
+    def at_xpath(self, xpath):
+        """
+        Sets the xpath for the last variation in the list.
+
+        Parameters
+        ----------
+        xpath : string
+            path expression to identify a node within the XML file
+            of the MuJoCo environment.
+        """
+        if self._list:
+            self._list[-1].xpath = xpath
+        return self
+
+    def attribute(self, attrib):
+        """
+        Sets the attribute for the last variation in the list.
+
+        Parameters
+        ----------
+        attrib : string
+            name of the dynamic parameter to randomize within the
+            node defined in xpath.
+        """
+        if self._list:
+            self._list[-1].attrib = attrib
+        return self
+
+    def with_method(self, method):
+        """
+        Sets the method to apply the random coefficient for the last variation
+        in the list.
+
+        Parameters
+        ----------
+        method : VariationMethod
+            if equal to "absolute", it sets the dynamic parameter
+            equal to the random coefficient obtained from the distribution, or
+            if equal to "coefficient", it multiplies the default value provided
+            in the XML file by the random coefficient.
+        """
+        if self._list:
+            self._list[-1].method = method
+        return self
+
+    def sampled_from(self, distribution):
+        """
+        Sets the distribution where the random coefficient is sampled from for
+        the last variation in the list.
+
+        Parameters
+        ----------
+        distribution : VariationDistribution
+            it specifies the probability distribution used to obtain the random
+            coefficient.
+        """
+        if self._list:
+            self._list[-1].distribution = distribution
+        return self
+
+    def with_range(self, low, high):
```

Comment:
what stops me from using this method along with `with_distribution()`?

Reply:
Two ideas to fix this. First one is to raise an error during the runtime. e.g.
```
variations = Variations()
variations.randomize() \
        .attribute("gear") \
        .at_xpath(".//motor[@name='a1']") \
        .with_method(Method.COEFFICIENT) \
        .sampled_from(Distribution.UNIFORM) \
        .with_mean_std(0.5, 1.5) \
        .add()
```

will raise
```
ValueError: Need to call with_range when sampled from Uniform
```
The current commit implements this idea.

Another idea is implement something like this
<img width="562" alt="screen shot 2018-06-06 at 12 17 03 pm" src="https://user-images.githubusercontent.com/8605658/41060145-8d5cb386-6983-11e8-9352-2f2746b438ef.png">

There is no `with_mean_std` when user calls `sampled_from(UNIFORM)`
I tried using mixin to achieve this but failed. Because using inheritance will return either VariationsGaussian or VariationsUniform in sampled_from. There's no way to tell which one should be called before running time.